dot/core: call core execute block

edwardmack opened the pull request:
<!--
 Before submitting a PR please ensure all code is commented and all tests are passing. Make sure to review the changes and ensure that only required changes are included (eg. no unnecessary reformatting by your editor)
-->

<!-- Please begin the title of this PR with a list of the packages touched (eg. "cmd, rpc: Add Literal Bells and Whistles") -->


<!-- Brief but specific list of changes made, describe the change in functionality rather than the change in code -->
## Changes
- Added call to Core_executeBlock in sync.go to handleBlockResponse

<!-- Details on how to run only the tests for the changes in the PR -->
## Tests:
```
 go test ./dot/core/... 
```

<!-- Issues that this PR will close -->
<!-- 
    NOTE: you must say 'closes #xx' or 'fixes #xx' for EACH issue this closes. 
    eg: 'closes #1 and closes #2'
    See: https://help.github.com/en/articles/closing-issues-using-keywords
-->
### Issues:
closes: #681

noot commented on dot/core/service.go:
```diff
@@ -602,6 +603,142 @@ func (s *Service) createBlockResponse(msg *network.BlockRequestMessage) (*networ
+			// if executeBlock return a non-empty byte array something when wrong
+			if !reflect.DeepEqual(res, []byte{}) {
+				log.Error("[core] execute block call failed", "err", res)
+			}
```
looks good, when you merge with development the logic for this function is moved to syncer.handleBlockResponse so you can move this there

edwardmack replied:
Any suggestion how I can get reference to runtime from within syncer.handleBlockResponse to call executeBlock?  Should I add reference to runtime in the syncer struct?  Would that be ugly because than syncer would require runtime?  Suggestions?

noot replied:
I think adding runtime to the syncer for now is okay, it would be nice to have all the runtime functions in one place but we can figure that out later

edwardmack replied:
ok

ryanchristo replied:
This was mention in #709 as well, moving it to syncer would be ok but all the runtime methods should be in one place. I will address in #709 or open a new issue if you do not address here.

noot commented on dot/core/service.go:
```diff
@@ -602,6 +603,142 @@ func (s *Service) createBlockResponse(msg *network.BlockRequestMessage) (*networ
+			}
+			// if executeBlock return a non-empty byte array something when wrong
+			if !reflect.DeepEqual(res, []byte{}) {
+				log.Error("[core] execute block call failed", "err", res)
```
we want to return if this fails

edwardmack replied:
ok

noot commented on lib/babe/build.go:
```diff
@@ -99,6 +100,34 @@ func (b *Session) buildBlock(parent *types.Header, slot Slot) (*types.Block, err
+	if !reflect.DeepEqual(res, []byte{}) {
+		return nil, errors.New("[babe] execute block failed")
+	}
+	log.Trace("[babe] execute block")
```
can remove this, we don't need to execute it after building. however if you could add this to the build block tests that would be good!

edwardmack replied:
removed this, and added test to build_test.go

noot commented on lib/runtime/imports.go:
```diff
@@ -248,7 +248,7 @@ func ext_storage_root(context unsafe.Pointer, resultPtr int32) {
 //export ext_storage_changes_root
 func ext_storage_changes_root(context unsafe.Pointer, a, b, c int32) int32 {
-	log.Trace("[ext_storage_changes_root] executing...")
+	log.Trace("[ext_storage_changes_root] executing...", "a", a, "b", b, "c", c)
```
don't need the paramters for now

edwardmack replied:
since this is a Trace log, I thought it would be useful to add information about values passed into the function.

noot replied:
okay that's fine

edwardmack commented:
I think I've address all comments, let me know. Thanks.

noot commented on dot/core/sync.go:
```diff
@@ -257,7 +263,33 @@ func (s *Syncer) handleBlockResponse(msg *network.BlockResponseMessage) (int64,
 
-			// TODO: execute block and verify authorship right
+			// prepare block for sending to core_executeBlock,
+			//  core_executeBlock fails if Digest and Body data are sent
```
it seems strange that it fails if the Body is sent, it needs the extrinsics to validate the block?

edwardmack replied:
I agree, should I try to determine why it doesn't work with extrinsics before merging this?

noot replied:
that's probably a good idea, what extrinsics are you using to test? because some of them (eg. transfer) won't work because the account has no balance

edwardmack replied:
I think I'm probably not using extrinsic correctly, I was using the array at line 124 in build_test, but I'm not using in correctly.  Do you have an example of an extrinsic I can try?

edwardmack replied:
I found my issue, I should have been passing encoded `types.BlockData` into execute block (I was using  `types.Block`).

noot commented on dot/core/sync.go:
```diff
@@ -257,7 +263,33 @@ func (s *Syncer) handleBlockResponse(msg *network.BlockResponseMessage) (int64,
+			bdEnc, err := blockData.Encode()
+			if err != nil {
+				return 0, err
+			}
```
could you move this logic to the `executeBlock` function, and have `executeBlock` take a block instead?

edwardmack replied:
done.

edwardmack commented:
@noot Please re-review, I think I've addressed all issues and now have 60% diff-coverage.  I addressed issue when calling executeBlock (I was passing in Block, when I should have been using BlockData).

edwardmack commented:
@noot The only way I can get core execute block to work is when I pass an encoded block that does not contain header digest or body.  I know this doesn't seem right, but the only combination of bytes that I could find to work was encode hash, encoded number, encode hash, encoded hash.  Add anything else seems to cause an error.

I wonder if we'll see different behavior when we update runtime?  Not sure if we should commit this, but it's passing tests now and two clients do sync without crashing.

noot commented on dot/core/sync_test.go:
```diff
@@ -49,6 +59,11 @@ func newTestSyncer(t *testing.T, cfg *SyncerConfig) *Syncer {
 
+	if cfg.Runtime == nil {
+		cfg.Runtime = runtime.NewTestRuntime(t, tests.POLKADOT_RUNTIME)
+		//cfg.Runtime = runtime.NewTestRuntime(t, tests.ED_RUNTIME)
```
remove

noot commented on dot/core/sync_test.go:
```diff
@@ -371,3 +386,90 @@ func TestWatchForResponses_MissingBlocks(t *testing.T) {
+	res, err := syncer.handleBlockResponse(msg)
+	require.Nil(t, err)
+
+	t.Logf("res %v", res)
```
remove
